purchases/src/main/kotlin/com/revenuecat/purchases/google/usecase/QueryProductDetailsUseCase.kt

```diff
@@ -0,0 +1,129 @@
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.StoreProductsCallback
+import com.revenuecat.purchases.common.between
+import com.revenuecat.purchases.common.diagnostics.DiagnosticsTracker
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.google.billingResponseToPurchasesError
+import com.revenuecat.purchases.google.buildQueryProductDetailsParams
+import com.revenuecat.purchases.google.toGoogleProductType
+import com.revenuecat.purchases.google.toHumanReadableDescription
+import com.revenuecat.purchases.google.toStoreProducts
+import com.revenuecat.purchases.strings.OfferingStrings
+import java.util.Date
+import kotlin.time.Duration
+
+internal data class QueryProductDetailsUseCaseParams(
+    val dateProvider: DateProvider = DefaultDateProvider(),
+    val diagnosticsTrackerIfEnabled: DiagnosticsTracker?,
+    val productIds: Set<String>,
+    val productType: ProductType,
+)
+
+internal class QueryProductDetailsUseCase(
+    private val useCaseParams: QueryProductDetailsUseCaseParams,
+    val onReceive: StoreProductsCallback,
+    val onError: PurchasesErrorCallback,
+    val withConnectedClient: (BillingClient.() -> Unit) -> Unit,
+    executeRequestOnUIThread: ((PurchasesError?) -> Unit) -> Unit,
+) : UseCase<List<ProductDetails>>(onError, executeRequestOnUIThread) {
+
+    override fun executeAsync() {
+        val nonEmptyProductIds = useCaseParams.productIds.filter { it.isNotEmpty() }.toSet()
+
+        if (nonEmptyProductIds.isEmpty()) {
+            log(LogIntent.DEBUG, OfferingStrings.EMPTY_PRODUCT_ID_LIST)
+            onReceive(emptyList())
+            return
+        }
+        withConnectedClient {
+            val googleType: String = useCaseParams.productType.toGoogleProductType() ?: BillingClient.ProductType.INAPP
+            val params = googleType.buildQueryProductDetailsParams(nonEmptyProductIds)
+
+            queryProductDetailsAsyncEnsuringOneResponse(
+                this,
+                googleType,
+                params,
+            ) { billingResult, productDetailsList ->
+                processResult(billingResult, productDetailsList)
+            }
+        }
+    }
+
+    override fun onOk(received: List<ProductDetails>) {
+        log(
+            LogIntent.DEBUG,
+            OfferingStrings.FETCHING_PRODUCTS_FINISHED.format(useCaseParams.productIds.joinToString()),
+        )
+        log(
+            LogIntent.PURCHASE,
+            OfferingStrings.RETRIEVED_PRODUCTS.format(received.joinToString { it.toString() }),
+        )
+        received.takeUnless { it.isEmpty() }?.forEach {
+            log(LogIntent.PURCHASE, OfferingStrings.LIST_PRODUCTS.format(it.productId, it))
+        }
+
+        val storeProducts = received.toStoreProducts()
+        onReceive(storeProducts)
+    }
+
+    override fun forwardError(billingResult: BillingResult, onError: PurchasesErrorCallback) {
```

Comment:
I'm wondering, would it be possible to move this to the `UseCase` base class? We could change the logs so we can print something generic and just provide a name to each use case so it can be printed in these logs... Lmk what you think!